### ISSUE TITLE
Hardcoded "Sec:" when referencing labels of sections

### ISSUE THREAD

**Issue by niazim3** [JacquesCarette/Drasil#971]
(Moved from #963)

Sometimes a section needs to be referenced in paragraphs of other sections. Since the document languages build sections sort of "in parallel", do we want to `makeRef` to the label that a section uses or ideally just `makeref` to the section itself?

In the meeting yesterday, it was mentioned that it seems fine semantically to `makeRef` directly to a label and hardcode the `Sec:` in https://github.com/JacquesCarette/Drasil/blob/0b501b02ff85404b20934f6d23308552f8e9c4b3/code/drasil-lang/Language/Drasil/Reference.hs#L251-L256.

@elwazana came across a situation, however, where this hack interferes with the SWHS implementation.

**Comment by JacquesCarette** [comment]
Yes, I need to think about this.  But I'll think about #968 first.

**Comment by elwazana** [comment]
@JacquesCarette Hello Dr. Carette, the situation I came across is in SWHS. The labels for the assumptions are referenced by the TMs directly, which causes them to have a `Sec:` which breaks the link. 

A possible solution to this is to add a filed into the Label record type, similar to how Changes are differentiated into Likely and Unlikely:
![image](https://user-images.githubusercontent.com/35268928/43536198-01d1d03e-958a-11e8-81a9-4cbcbec19a91.png)

**Comment by niazim3** [comment]
I was thinking of implementing something like the following:
```Haskell
-- | For building references. Defines the possible type of reference.
data RefType = Tab    -- ^ Table
             | Fig    -- ^ Figure
             | Sect   -- ^ Section
             | Def DType  -- ^ Definition (includes theoretical models) (DType used to set shortnames)
             | Mod    -- ^ Module
            ...
             | Lbl RefType -- ^ Label


instance Show RefType where
  show Tab    = "Table"
  ...
  show (Lbl x)= show x
  show Mod    = "Module"
  ...
```
, but there isn't really a way to pass in the `RefType` argument into 
```Haskell
--FIXME: assumes reference to a direct Label is for a section
--Should refer to an object WITH a variable.
--Can be removed once sections have labels.
instance Referable Label where
  refAdd lb = "Sec:" ++ (getAdd (lb ^. getRefAdd))
  rType _ = Lbl (rType x) 
```

Since `Label`s are not associated with different chunk types.

I think for now, the `missingP` hack should be reintroduced for referencing sections to remove the hardcoded "Sec:", and replace the hardcoded "Sec:" to a hardcoded "A:" since SWHS/NoPCM are the only examples that reference a label (that is not related to a section) directly.

Or should the broken referencing be pushed to master, with this issue remaining open?

@JacquesCarette @szymczdm

**Comment by JacquesCarette** [comment]
What we should be doing is having label-building combinators which take a `RefType` as input. By the time you're at `Referable`, it's too late.

**Comment by niazim3** [comment]
That would mean that `Label`s will be associated with a specific chunk type, correct? Do we want that? I'm asking because in https://github.com/JacquesCarette/Drasil/issues/567 it was mentioned that Label is not attached to anything.

**Comment by JacquesCarette** [comment]
Yes, it means that a given label will be associated with a particular kind of object. Labels are already implicitly associated to particular objects via their name.

From a technical standpoint, it is true, and by design, that labels are not attached to anything in particular. Labels, as a data-structure, do not know what they are attached to. But the creator of labels does know what they will attach a *particular* label to, and can use that information, when a particular label is being created.

**Comment by JacquesCarette** [comment]
A sort-of design has been added to the discussion at 371f46b.

**Comment by niazim3** [comment]
Just noting here that by removing the `RefType` field from `Label` as outlined in the design comment @JacquesCarette mentions above, there is no way to know what `RefType` needs to be set, since it used to simply be 
https://github.com/JacquesCarette/Drasil/blob/9870e0bebb30647857f2d93e49a68fab63e475f2/code/drasil-lang/Language/Drasil/Reference.hs#L267
and now it's
```Haskell
--Should refer to an object WITH a variable.
--Can be removed once sections have labels.
instance Referable Label where
  refAdd lb@(Lbl _ _ _) = getAdd (lb ^. getRefAdd)
  rType  (Lbl _ _ _) = Label --FIXME: need to know the RefType somehow!
```
.
SInce this is the case, I believe this issue needs to be looked into in more detail. For the sake of merging the addLabels branch in as soon as possible, the hardcoded section that comes from referencing a label directly will be updated in stable to remove the discrepancies in the logs. These are the only broken references in Drasil as of commit ____.

**Comment by elwazana** [comment]
As it is now the `RefType` being fed into the constructor for `Label` also adds a prefix to the `RefAdd` itself. 
Would a solution to this issue then be implementing a parser that takes the hardcoded `RefAdd` prefix and outputting the associated `RefType`? This way actually referencing the label will display the correct `RefType`s corresponding prefix, without needing to have a `RefType` field in `Label`.

@JacquesCarette Any thoughts?

**Comment by szymczdm** [comment]
I don't think using a hard-coded parser would be the right design. As for what to do instead, I'm not sure yet, but I'll think about it.

**Comment by JacquesCarette** [comment]
I definitely agree that a parser is the wrong design.

The design was supposed to be that you create labels with all the necessary information for building all the pieces of the label (i.e. reference address, short name) at that time. In many ways, I think such a design ought to make `Referable` obsolete. It tries to reconstruct some information 'too late'.
